<div style="text-align:left;font-size:2em"><span style="font-weight:bolder;font-size:1.25em">SP2273 | Learning Portfolio</span><br><br><span style="font-weight:bold;color:darkred">Storing Data (Need) Exercises</span></div>

In [ ]:
!pip install numpy # if needed

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.6/6.6 MB 30.2 MB/s  0:00:00 eta 0:00:01


---
# Exercise 1

Similarities:
1. Both can be indexed.

2. Both are mutable.

Differences:
1. List elements can be of different types, while array elements cannot be.

2. Operations can be done on all elements of an array by doing the operation on the array (eg. `array_name * 2` multiplies all elements by 2)

A dictionary is a two-dimensional way of storing data with a paired key/value.


---
# Exercise 2
Original code:


In [ ]:
# Original code:
py_list = ["a1", "b2", "c3", "d4", "e5", "f6", "g7", "h8", "i9", "j10"]
                    # Prints 'a1'
                    # Prints 'c3'
                    # Prints 'e5'
                    # Prints 'g7'
print(py_list[8])   # Prints 'i9'

Fixed code:

In [ ]:
# changed code no. 1
py_list = ["a1", "b2", "c3", "d4", "e5", "f6", "g7", "h8", "i9", "j10"]
print(py_list[0])   # Prints 'a1'
print(py_list[2])   # Prints 'c3'
print(py_list[4])   # Prints 'e5'
print(py_list[6])   # Prints 'g7'
print(py_list[8])   # Prints 'i9'

a1
c3
e5
g7
i9


In [ ]:
# Changed code no. 2
# my bad i didn't read instructions properly
py_list = ["a1", "b2", "c3", "d4", "e5", "f6", "g7", "h8", "i9", "j10"]

for i in py_list:
    num = float(i[-1:]) # take last character of string
    if num%2 == 1: # if number odd
        print(i) # print it

a1
c3
e5
g7
i9


---
# Exercise 3

In [ ]:
elements = ['Hydrogen',
            'Helium', 'Lithium',
            'Beryllium', 'Boron', 'Carbon',
            'Nitrogen', 'Oxygen',
            'Fluorine',
            'Neon']
# Forward indexing
print(elements[3]) # python starts from 0 as first element

# Backward indexing
print(elements[-7]) # 7th from the back, as back starts from -1

Beryllium
Beryllium


---
# Exercise 4

In [ ]:
import numpy as np
numbers= np.array([45, 60, 1, 30, 96, 1, 96, 57, 16, 1,
        99, 62, 86, 43, 42, 60, 59, 1, 1, 35,
        83, 47, 34, 28, 68, 23, 22, 92, 1, 79,
        1, 29, 94, 72, 46, 47, 1, 74, 32, 20,
        8, 37, 35, 1, 89, 29, 86, 19, 43, 61])

print(np.count_nonzero(numbers == 1))
    

9


---
# Exercise 5

In [ ]:
import numpy as np

# made this a function just for neatness
def manual_r(array1: np.array, array2: np.array): 
    '''Manually calculate r.'''
    array1_mean = np.mean(array1) # mean of array 1
    array1_minusmean = array1 - array1_mean # making x_i - x_mean, for manual r
    array1_std = np.std(array1) # finding standard deviation

    array2_mean = np.mean(array2) # mean of array 2
    array2_minusmean = array2 - array2_mean # making x_i - x_mean, for manual r
    array2_std = np.std(array2) # finding standard deviation

    # formula for manual r
    r = (np.sum(
        np.multiply(
            array1_minusmean, array2_minusmean)))/(
                                                    array1_std * array2_std)
    
    return r

# Means and standard deviations
x_1 = np.array([1, 4, 9, 16, 25, 36, 49, 64, 81, 100])
print(f'x_1 mean = {np.mean(x_1)}, standard deviation = {np.std(x_1)}')

x_2 = np.array([1, 8, 27, 64, 125, 216, 343, 512, 729, 1000])
print(f'x_2 mean = {np.mean(x_2)}, standard deviation = {np.std(x_2)}')

# Manual r
r = manual_r(x_1, x_2)
print(f'Manually calculated r is {r}')

# np.corrcoef()
print(f'np.corrcoef() r value: {np.corrcoef(x_1, x_2)}')

x_1 mean = 38.5, standard deviation = 32.41990129534635
x_2 mean = 302.5, standard deviation = 326.0893282522444
Manually calculated r is 9.871796626270703
np.corrcoef() r value: [[1.         0.98717966]
 [0.98717966 1.        ]]


## Output:
x_1 mean = 38.5, standard deviation = 32.41990129534635 \
x_2 mean = 302.5, standard deviation = 326.0893282522444 \
Manually calculated r is 9.871796626270703 \
np.corrcoef() r value: [[1.         0.98717966] \
 [0.98717966 1.        ]]

 Manually calculating *r* finds the correlation coefficient, while 
 `np.corrcoef()` returns a matrix with the correlation coefficient for each
 factor.